src/lib.rs

```diff
@@ -1,718 +1,923 @@
-            if input.len() > usize {
-                buffer[..usize].copy_from_slice(&input[1..=usize]);
+struct Buffer<R>
+where
+    R: Read,
+{
+    reader: R,
+    buffer: BitVec,
+    ignore_eof: bool,
+}
 
-                Ok(Packet::Instrumentation(Instrumentation {
-                    buffer,
-                    size,
-                    port,
-                }))
-            } else {
-                // need more bytes
-                Err(Either::Right(NeedMoreBytes))
-            }
+impl<R> Buffer<R>
+where
+    R: Read,
+{
+    pub fn new(reader: R, ignore_eof: bool) -> Buffer<R> {
+        Buffer {
+            reader,
+            ignore_eof,
+            buffer: BitVec::new(),
         }
+    }
 
-        Header::LTS1 { tc } => {
-            // parse the payload
-            let mut delta = 0;
-            let mut cursor = 1u8;
-            loop {
-                let payload = input
-                    .get(usize::from(cursor))
-                    .cloned()
-                    .ok_or(Either::Right(NeedMoreBytes))?;
-
-                delta += (u32::from(payload) & 0b0111_1111) << (7 * (cursor - 1));
-
-                if payload & 0b1000_0000 == 0 {
-                    // the C (Continue) bit is zero; end of the packet
-                    break;
-                } else {
-                    // the C (Continue) bit is set
-
-                    if cursor == 4 {
-                        // payloads are at most 4 bytes in size
-                        // according to the spec the last C bit should be zero so this is an error
-
-                        // the final payload byte may have been lost and this could be a new
-                        // header byte so we consider the malformed packet to end at the third
-                        // payload byte
-                        return Err(Either::Left(Error::MalformedPacket {
-                            header,
-                            len: cursor,
-                        }));
-                    } else {
-                        cursor += 1;
-                        continue;
-                    }
+    /// Tries to read up to 32 bytes from [Self::reader]. Continuously retries if [ignore_eof] is set.
+    fn buffer_some(&mut self) -> Result<(), DecoderErrorInt> {
+        // `Read::read` reportedly reads in 32-byte chunks. Source:
+        // <https://github.com/rust-embedded/itm/blob/3e4251b42aa2e4b05ae372c47c7b835b8acae6dc/src/lib.rs#L42>.
+        let mut buffer: [u8; 32] = [0; 32];
+        match self.reader.read(&mut buffer) {
+            Ok(0) => {
+                if self.ignore_eof {
+                    return self.buffer_some();
                 }
+                Err(DecoderErrorInt::Eof)
             }
+            Ok(n) => {
+                let mut bv = BitVec::<LocalBits, _>::from_vec(buffer[0..n].to_vec());
+                bv.reverse();
+                bv.append(&mut self.buffer);
+                self.buffer.append(&mut bv);
```

Comment:
This function is the most resource-intensive. Particularly the reversal. The code if from before `itm-decode` was refactored, and can likely be optimized.